src/groupByKey.ts

```diff
@@ -1,7 +1,27 @@
 'use strict';
 
-// describe interfaces
+interface Item {
+  [key: string]: any
+}
+
+interface ItemsGroup<T> extends Item{
+  [key: string]: T[]
+}
+
+export function groupByKey<T, K extends keyof T>(
+  array: Array<T>, key: K): ItemsGroup<T> {
+  const result: ItemsGroup<T> = {};
+
+  for (const obj of array) {
+    const resultKey = obj[key];
+
+    if ((typeof resultKey === 'string' || typeof resultKey === 'number')
+      && resultKey in result) {
+      result[resultKey].push(obj);
+    } else if (typeof resultKey === 'string' || typeof resultKey === 'number') {
+      result[resultKey] = [obj];
+    }
+  }
```

Comment:
use reduce method to simplify your solution

Reply:
This way you can easily extract object from an array and make your if statement more simple